### PR TITLE
Added OTA feature and example

### PR DESCRIPTION
**Pull request by hallard**
Added OTA feature
Created according new example `SimpleMQTTClientOTA`

### CODE REVIEW

File: examples/SimpleMQTTClientOTA/SimpleMQTTClientOTA.ino [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  client.enableHTTPWebUpdater(); // Enable the web updater. User and password default to values of MQTTUsername and MQTTPassword. These can be overrited with enableHTTPWebUpdater("user", "password").
+  client.enableOTA(); // Enable OTA here, OTA networkk name will be MQTT client hostname (so hostname here)
+  //client.enableOTA("my_ota_password"); // Enable OTA with password
+  //client.enableOTA("my_ota_password"; 32000); // Enable OTA with password and port
+
+  // You can activate the retain flag by setting the third parameter to true
+  client.enableLastWillMessage(topicbase+ "/LWT", "Offline", true);  
+
+  // This is the $1 000 0000 question should we leave this here to be able to manage 
+  // ourselve or do we need to to this in class adding callback if needed ?
```
Comment (EdJoPaTo) [inline examples/SimpleMQTTClientOTA/SimpleMQTTClientOTA.ino:60]: Personally: I use EspMqttClient to not do stuff like this myself.
Just include it within the library itself and provide a useful default.
If I want to do it myself I can just not use `client.enableOTA()` and it myself.

Comment (EdJoPaTo) [reply examples/SimpleMQTTClientOTA/SimpleMQTTClientOTA.ino:60]: Also: As this is the main difference between the normal and the OTA example: add the `client.enableOTA()` thingies to the main example to keep it simpler.

File: src/EspMQTTClient.cpp [hunk 2]
```diff
@@ -131,6 +154,11 @@ void EspMQTTClient::enableLastWillMessage(const char* topic, const char* message
   _mqttLastWillRetain = retain;
 }
 
+void EspMQTTClient::enableLastWillMessage(const String &topic, const String &message, const bool retain )
```
Comment (EdJoPaTo) [inline src/EspMQTTClient.cpp:157]: I think this method is unrelated to this PR? Also as every method currently expects `const char*` it should be that way for everything.

I would remove it

File: src/EspMQTTClient.h [hunk 3]
```diff
@@ -157,6 +165,8 @@ class EspMQTTClient
   inline const char* getMqttClientName() { return _mqttClientName; };
   inline const char* getMqttServerIp() { return _mqttServerIp; };
   inline const short getMqttServerPort() { return _mqttServerPort; };
+  inline void disconnectMqtt() { _mqttClient.disconnect(); };
```
Comment (EdJoPaTo) [inline src/EspMQTTClient.h:168]: Not sure if this is really needed? looks like you tested with it in the example code but stopped using it. Maybe also get rid of this addition then?

**Review by EdJoPaTo — COMMENTED** [review]
As I implemented OTA myself lately and then stumbled upon this PR, I tested it. The default unprotected OTA works for me.

Personally I would like to see some things changed but its not my repo ;)

**Comment by plapointe6** [comment]
@EdJoPaTo made a really good review of these changes that worth being adressed. Does ArduinoOTA is a third party library ? If yes, this PR will add a dependency. Also, did you tested your changes on an ESP32 and an ESP8266 ? Thanks for your contribution !

**Comment by hallard** [comment]
Hi Guys
Thanks for your review, ArduinoOTA is included in core repo for Arduino ESP8266 and ESP32 so don't need to install anything as soon you use this EspMQTTClient from Arduino IDE and it should be the same for platformIO also (not tested this one) but I tested this sample code on ESP32 and ESP8266 of course ;-)

**Comment by EdJoPaTo** [comment]
Just saw @hallard already responded, so I just add the fitting link here
https://github.com/esp8266/Arduino/tree/master/libraries/ArduinoOTA

File: src/EspMQTTClient.h (hunk 3 above)
Comment (hallard) [reply src/EspMQTTClient.h:168]: Yes I used it because long time ago OTA while working with other connected TCP was difficult to go until the end without errors, but looks like now it correctly handed by the framework. But does not hurt to leave it, if some have issues during OTA it's the first thing to do, close all other network connectivity when starting OTA

**Comment by EdJoPaTo** [comment]
Personally, I tested this with ESP8266 and PlatformIO. I also used ArduinoOTA in the exact same places outside of the lib from the Arduino IDE (only difference: I called `ArduinoOTA.begin();` in `onConnectionEstablished` and not `onWiFiConnectionEstablished` as I was outside of the lib. As it worked with Platform IO I do not assume that makes a big difference)

I can check if it works with an ESP32 in combination with PlatformIO later.

**Comment by plapointe6** [comment]
Thanks guys for your great work. I will merge that soon because you already implemented and tested everything. I am a bit late but I was not at home the last three weeks.

**Comment by EdJoPaTo** [comment]
I would like to use this from the lib and am curious what the current status is. There are still some review things open which I think should be addressed. @hallard any progress on that or wasn't there time for that?

Just wanna make sure this wasn't just forgotten :)

**Comment by pr0glor** [comment]
Is this repo dead or is there any progress to be expected in the near future?

**Comment by EdJoPaTo** [comment]
This repo isnt dead, just this PR sadly seems to be… I guess someone could pick up the work and the suggestions and provide a new PR?
